internal/datastore/mysql/datastore.go

```diff
@@ -0,0 +1,648 @@
+	colTimestamp        = "timestamp"
+	colNamespace        = "namespace"
+	colConfig           = "serialized_config"
+	colCreatedTxn       = "created_transaction"
+	colDeletedTxn       = "deleted_transaction"
+	colObjectID         = "object_id"
+	colRelation         = "relation"
+	colUsersetNamespace = "userset_namespace"
+	colUsersetObjectID  = "userset_object_id"
+	colUsersetRelation  = "userset_relation"
+
+	errRevision             = "unable to find revision: %w"
+	errCheckRevision        = "unable to check revision: %w"
+	errUnableToInstantiate  = "unable to instantiate datastore: %w"
+	errUnableToQueryTuples  = "unable to query tuples: %w"
+	errUnableToWriteTuples  = "unable to write tuples: %w"
+	errUnableToDeleteTuples = "unable to delete tuples: %w"
+	liveDeletedTxnID        = uint64(9223372036854775807)
+	batchDeleteSize         = 1000
+)
+
+var (
+	tracer = otel.Tracer("spicedb/internal/datastore/mysql")
+
+	// Unless specified otherwise, Go's MySQL driver will assume
+	// the server sends datetime in UTC,
+	// see https://github.com/go-sql-driver/mysql#loc. This parameter
+	// is unrelated to the session's timezone.
+	// If the server's global timezone is set to something other than UTC,
+	// the driver will incorrectly convert SELECT NOW(), because
+	// the default session timezone is the one specified by the server.
+	getNow = sb.Select("UTC_TIMESTAMP(6)")
+
+	sb = sq.StatementBuilder.PlaceholderFormat(sq.Question)
+)
+
+type sqlFilter interface {
+	ToSql() (string, []interface{}, error)
+}
+
+func NewMySQLDatastore(url string, options ...Option) (*Datastore, error) {
+	config, err := generateConfig(options)
+	if err != nil {
+		return nil, fmt.Errorf(errUnableToInstantiate, err)
+	}
+	connector, err := mysql.MySQLDriver{}.OpenConnector(url)
+	if err != nil {
+		return nil, fmt.Errorf("NewMySQLDatastore: failed to create connector: %w", err)
+	}
+	var db *sql.DB
+	if config.enablePrometheusStats {
+		connector, err = instrumentConnector(connector)
+		if err != nil {
+			return nil, fmt.Errorf("NewMySQLDatastore: unable to instrument connector: %w", err)
+		}
+		db = sql.OpenDB(connector)
+		// Create a new collector, the name will be used as a label on the metrics
+		collector := sqlstats.NewStatsCollector("spicedb", db)
+		err := prometheus.Register(collector)
+		if err != nil {
+			return nil, fmt.Errorf(errUnableToInstantiate, err)
+		}
+	} else {
+		db = sql.OpenDB(connector)
+	}
+	db.SetConnMaxLifetime(config.connMaxLifetime)
+	db.SetConnMaxIdleTime(config.connMaxIdleTime)
+	db.SetMaxOpenConns(config.maxOpenConns)
+	db.SetMaxIdleConns(config.maxOpenConns)
+
+	driver := migrations.NewMysqlDriverFromDB(db, config.tablePrefix)
+	queryBuilder := NewQueryBuilder(driver)
+
+	createTxn, _, err := sb.Insert(driver.RelationTupleTransaction()).Values().ToSql()
+	if err != nil {
+		return nil, fmt.Errorf("NewMySQLDatastore: %w", err)
+	}
+
+	// used for seeding the initial relation_tuple_transaction. using INSERT IGNORE on a known
```

Comment:
why is this created here and not in the seed method?

Reply:
Because the SQL are dynamically generated to support the table prefix functionality, it's generated once and kept as state. The datastore initialization seems like an appropriate place to set the state of the datastore structure.

An alternative is to extend `QueryBuilder` to also hold queries that do not need a builder. Another alternative is generating this string each time as part of the `IsReady()` method. That seemed unnecessarily work to repeat, but given `IsReady()` is not a hot path, I don't have strong opinions against putting it there.